File: October Challenge/Week2/C++/Day_9/SerializeandDeserializeBST.cpp
class Codec {

	public:

	// Encodes a tree to a single string.

	void helper(TreeNode* root, string &ans)
	{
		if(root==NULL)
    {
        return;
    }
    
    ans+=to_string(root->val);
    ans+='#';
    
    helper(root->left,ans);
    helper(root->right,ans);
    
    return;
	
	}

	string serialize(TreeNode* root) {
    
    string ans="";
    
    helper(root,ans);
    
    return ans;
    
}

	// Decodes your encoded data to tree.

	TreeNode* helper2(TreeNode* root,int value)
	{
		if(root==NULL)
		{
			TreeNode* temp=new TreeNode(value);
			return temp;
		}
    
		if(value<=root->val)
		{
			root->left=helper2(root->left,value);
		}
    
		else
		{
			root->right=helper2(root->right,value);
		}
    
		return root;
	}


	TreeNode* deserialize(string data) {
    
    int i=0;
    TreeNode* root=NULL;
    while(i<data.length())
    {
        string temp="";
        
        while(i<data.length() and data[i]!='#')
        {
            temp+=data[i];
            i++;
        }
        
        i++;
        
        root=helper2(root,stoi(temp));
    }
    
    return root;
    
}
};

File: October Challenge/Week2/Java/Day_9/SerializeAndDeserializeBST.java
// Problem : https://leetcode.com/problems/serialize-and-deserialize-bst/
// Solution:

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        recurS(root, sb);
        return sb.toString();
    }
    
    private void recurS(TreeNode node, StringBuilder sb) {
        if (node == null) {
            return;
        }
        sb.append(node.val);
        sb.append(',');
        recurS(node.left, sb);
        recurS(node.right, sb);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.length() <= 0) {
            return null;
        }
        Queue<String> queue = new LinkedList<>(Arrays.asList(data.split(",")));
        return recurD(queue, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    
    private TreeNode recurD(Queue<String> queue, int min, int max) {
        if (queue.size() <= 0) {
            return null;
        }
        int val = Integer.parseInt(queue.peek());
        if (val <=  min || val >= max) {
            return null;
        }
        queue.poll();
        TreeNode node = new TreeNode(val);
        node.left = recurD(queue, min, val);
        node.right = recurD(queue, val, max);
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// String tree = ser.serialize(root);
// TreeNode ans = deser.deserialize(tree);
// return ans;